Possibility for overriding fetch manually for JONSWAP initialization

MHBalsmeier opened the pull request:
# Pull Request Summary
This pull request adds the possibility to set the fetch that is used by the fetch-limited JONSWAP initialization manually in ww3_strt.inp.

Disclaimer: Not fully tested yet and not fully certain if this feature makes sense. If it does not make sense or should be implemented differently, feel free to close it or request changes.

## Description
The problem I am trying to solve here is described in #1445. In short, assuming the fetch to be the grid resolution is unphysical since the wave spectrum (and thus the wave height) as a function of the wind is a physical quantity that not should not depend on the grid size. Correct me if I am wrong.

### Issue(s) addressed
- fixes #1445 

### Commit Message
adding the possibility to set the fetch that is used by the fetch-limited JONSWAP initialization manually in ww3_strt.inp.
 
### Testing

* How were these changes tested? Locally for a 1 km real-data configuration and with some of the regression tests (only checked if they completed).
* Are the changes covered by regression tests? Not fully certain on that.
* Have the matrix regression tests been run? no
* Please indicate the expected changes in the regression test output: everything should be identical as before as long as the fetch-limited JONSWAP initialization isn't used.

MHBalsmeier committed:
FETCH parameter can be set in ww3_start with ITYPE = 3.

MHBalsmeier committed:
JONSWAP fetch parameter

The manually set JONSWAP fetch parameter is now part of the GRID type, included in the restart file and used by the subroutine W3UINI.

MHBalsmeier committed:
fix regarding latest commit: 100 was hard-coded fetch value

MHBalsmeier committed:
bugfix regarding manually set JONSWAP fetch on LL grid

MHBalsmeier committed:
JONSWAP fetch initialization: restart file format unchanged if TYPE is 'CALM'

JessicaMeixner-NOAA commented:
@MHBalsmeier  These changes are not covered by a regression test unless you have updated a regression test to use your new feature.  This should be included with your PR to make sure your feature is actually tested and stays working moving forward. 

To check to see if your code changes change answers, you would run the regression tests with develop and then with your branch and compare to see if things change.

MHBalsmeier committed:
added two new input directories to the mww3_test_08 regression test to account for the JONSWAP initialization with manually set fetch

MHBalsmeier committed:
superfluous write statement removed in w3iorsmd, new run of mww3_test_08 included in matrix.base

MHBalsmeier committed:
fetch read from restart file for wind initialization only if restart file exists

MHBalsmeier committed:
modifications in several ww3_strt.inp files: added 0 to indicate default fetch for JONSWAP initialization

MHBalsmeier committed:
manual fetch for new mww3_test_08 test case configs set to 10,000 m

MHBalsmeier committed:
improvement in w3iorsmd.F regarding manual fetch: when init type 'WIND' is the fallback, fetch is set to zero (default)

MHBalsmeier commented:
@JessicaMeixner-NOAA 

* I included two new input directories in mww3_test_08 that test my new feature. In the two new directories the fetch is set to 100 km. One of them is included in the matrix.base file.
* For the already existing test cases that use the JONSWAP initialization I added a line in ww3_strt.inp to make sure they still run and deliver the same results as before.
* I modified the module w3iorsmd.F90 compared to my first version: when the 'WIND' initialization type is set as the fallback because no readable restart file is found (label 800), it no longer tries to read the fetch from the restart file but sets the fetch to zero.
* Regarding the regression tests: I do not have access to a HPC infrastructure and none of the regtests/bin/matrix_cmake_* files worked for me. I therefore ran a subset of the tests. The results are attached.

[matrixCompSummary.txt](https://github.com/user-attachments/files/20653555/matrixCompSummary.txt)
[matrixCompFull.txt](https://github.com/user-attachments/files/20653556/matrixCompFull.txt)

JessicaMeixner-NOAA commented:
@MHBalsmeier thank you for running the regtests that you were able to run and adding a regression test.  I'll review this PR again in the next day or two and let you know if I have any comments/questions. I'll run the full set of regtests on my end as well.

JessicaMeixner-NOAA commented:
@MHBalsmeier  - I have run the regression tests on my end and here are the results: 

```
**********************************************************************
********************* non-identical cases ****************************
**********************************************************************
mww3_test_03/./work_PR1_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UQ_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UNO_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (16 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (9 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (16 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (15 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (16 files differ)
mww3_test_03/./work_PR2_UQ_MPI_d2                     (16 files differ)
mww3_test_03/./work_PR3_UQ_MPI_e                     (1 files differ)
mww3_test_03/./work_PR3_UNO_MPI_e_c                     (1 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (15 files differ)
mww3_test_08/./work_lowres_jonswap                     (directory not found)
mww3_test_09/./work_MPI_ASCII                     (0 files differ)
ww3_tp2.10/./work_MPI_OMPH                     (7 files differ)
ww3_tp2.15/./work_PR3_UQ_RHO                     (1 files differ)
ww3_tp2.15/./work_5km                     (1 files differ)
ww3_tp2.15/./work_PR3_UQ_RHO_MPI                     (1 files differ)
ww3_tp2.15/./work_ST6FLX5                     (1 files differ)
ww3_tp2.15/./work_PR3_UQ                     (1 files differ)
ww3_tp2.15/./work_MPI_5km                     (1 files differ)
ww3_tp2.15/./work_PR3_UQ_MPI                     (1 files differ)
ww3_tp2.15/./work_PR3_UQ_CONST                     (1 files differ)
ww3_tp2.15/./work_ST4FLX5                     (1 files differ)
ww3_tp2.16/./work_MPI_OMPH                     (4 files differ)
ww3_tp2.21/./work_ma                     (1 files differ)
ww3_tp2.21/./work_b_metis                     (1 files differ)
ww3_tp2.21/./work_a                     (1 files differ)
ww3_tp2.21/./work_mb                     (1 files differ)
ww3_tp2.21/./work_b                     (1 files differ)
ww3_tp2.6/./work_ST4_ASCII                     (0 files differ)
ww3_ufs1.3/./work_a                     (3 files differ)
```
We do have some tests and are known to be non b4b but several tests beyond what I was originally expecting have diffs: 

```
ww3_tp2.10/./work_MPI_OMPH                     (7 files differ)
ww3_tp2.15/./work_PR3_UQ_RHO                     (1 files differ)
ww3_tp2.15/./work_5km                     (1 files differ)
ww3_tp2.15/./work_PR3_UQ_RHO_MPI                     (1 files differ)
ww3_tp2.15/./work_ST6FLX5                     (1 files differ)
ww3_tp2.15/./work_PR3_UQ                     (1 files differ)
ww3_tp2.15/./work_MPI_5km                     (1 files differ)
ww3_tp2.15/./work_PR3_UQ_MPI                     (1 files differ)
ww3_tp2.15/./work_PR3_UQ_CONST                     (1 files differ)
ww3_tp2.15/./work_ST4FLX5                     (1 files differ)
ww3_tp2.21/./work_ma                     (1 files differ)
ww3_tp2.21/./work_b_metis                     (1 files differ)
ww3_tp2.21/./work_a                     (1 files differ)
ww3_tp2.21/./work_mb                     (1 files differ)
ww3_tp2.21/./work_b                     (1 files differ)
```

These are all differing in the restart files.  Technically the code changes here require exiting options to add things to the restart, which means we should be updating the restart version number, which is very disruptive. Moreover, existing options seem to require this new variable.  

I have a few suggestions for a path forward: 

1. No code changes, and run a longer spin-up.  When starting from rest, we typically expect at least 2 weeks of spin-up and sometimes go up to 1 month before we would start looking at output of a simulation.   This should resolve your intial low values issue and you'd want to spin-up swell anyways. 

2. Instead of adding to an existing ITYPE = 3, create a new itype which would mean existing regression tests would not need to change answer, nor would you need to change input files of tests you were not intentionally changing to test this feature and we would not have to change the version number of the restart.

MHBalsmeier committed:
changed approach to manual fetch in JONSWAP spectrum

new ITYPE=6 for manual fetch feature

new init type FTCH in w3iorsmd

modified previously changed ww3_strt.inp files back to their previous versions

MHBalsmeier committed:
modified ww3_strt.inp files of tests mww3_test_08/i_lowres_*_jonswap to conform with new definition of ITYPE = 6, fetch set to 10 km

MHBalsmeier committed:
itype = 6 was not allowed yet, ww3_strt.inp files of new JONSWAP test cases were wrong

MHBalsmeier committed:
improvement regarding manually setting the fetch

restart file was not properly read and written

MHBalsmeier commented:
@JessicaMeixner-NOAA Thank you for the feedback. I really do need this feature since I have a use case where a longer spin-up is not possible. I now found the time to modify my implementation:

* implementation of a new `itype = 6`
* changed ww3_strt.inp files of tests that I did not modify back to their previous versions
* The communication from ww3_strt to ww3_shel/ww3_multi goes via the restart files. Therefore I had to do some changes to w3iorsmd. A new `TYPE = 'FETCH'` is now implemented there to leave the 'WIND' case unchanged. I hope this does not make a change to the version number of the restart files necessary.

As above, here again are the results of the limited subset of regression tests that I am able to run.
[matrixCompSummary.txt](https://github.com/user-attachments/files/20967894/matrixCompSummary.txt)
[matrixCompFull.txt](https://github.com/user-attachments/files/20967895/matrixCompFull.txt)

JessicaMeixner-NOAA commented:
Thanks @MHBalsmeier 

I've quickly gone over the code changes and on first glance they look okay to me.  One quick question I do have is if you could help me understand why the 3 additional new tests?  Are they each using this new feature or are there other aspects of them?   Can you provide a little more information about each of them as an overview?

MHBalsmeier committed:
modified ww3_strt.inp back to the previous version (without the fetch parameter since it is itype=3)

MHBalsmeier commented:
@JessicaMeixner-NOAA Sure, regarding the tests:

* The modification of regtests/ww3_tp2.15/input_const/ww3_strt.inp was unintended, I had modified it before and did not set it back to the old version (it is itype = 3 so everything should stay as it is).
* Regarding the other two: I went through the [list of regression tests](https://github.com/NOAA-EMC/WW3/wiki/Developer-Guide#regression-testing-in-wavewatch-iii) and thought that mww3_test_08 was the most suitable for this feature. i_lowres_multi_jonswap is a copy of i_lowres_multi and i_lowres_shel_jonswap is a copy of i_lowres_shel, both times with only the ww3_strt.inp file modified to itype = 6 and a manual fetch of 10 km. The reason for two tests is that it is probably useful to cover both the single- and multi-grid case.

MHBalsmeier commented:
@JessicaMeixner-NOAA Is there anything more I can do to get this merged?

JessicaMeixner-NOAA commented:
@MHBalsmeier  - Apologies this got lost in leave. We are trying to update our libraries for the machines where we do some testing, I'll need to get that updated and then I can come back to looking at this.

JessicaMeixner-NOAA commented on model/src/w3iorsmd.F90:
```diff
@@ -619,7 +621,11 @@ SUBROUTINE W3IORS ( INXOUT, NDSR, DUMFPI, IMOD, FLRSTRT )
-      IF ( TYPE.EQ.'WIND' .OR. TYPE.EQ.'CALM' ) THEN
+      IF ( TYPE.EQ.'FTCH' ) THEN
+        RPOS  = 1_8 + LRECL*(3-1_8)
+        WRITE (NDSR,POS=RPOS,ERR=803,IOSTAT=IERR) FETCH
```
@MHBalsmeier    After a recent PR we need to update this to instead be: 

          WRITE (NDSR,POS=RPOS,IOSTAT=IERR) FETCH
            IF (IERR.NE.0) CALL EXTIOF(NDSE,IERR,'W3IORS','',31, &
                                       ISWRITE=.TRUE.,POS=RPOS)

JessicaMeixner-NOAA commented on model/src/w3iorsmd.F90:
```diff
@@ -741,7 +747,13 @@ SUBROUTINE W3IORS ( INXOUT, NDSR, DUMFPI, IMOD, FLRSTRT )
-      IF ( TYPE.EQ.'WIND' .OR. TYPE.EQ.'CALM' ) THEN
+      IF ( TYPE.EQ.'FTCH') THEN
+        RPOS  = 1_8 + LRECL*(3-1_8)
+        READ(NDSR, POS=RPOS,ERR=802,IOSTAT=IERR) FETCH
```
After a recent PR we need to update this to be: 

            READ(NDSR, POS=RPOS,IOSTAT=IERR) FETCH
            IF (IERR.NE.0) CALL EXTIOF(NDSE,IERR,'W3IORS','',30)

JessicaMeixner-NOAA commented on model/src/ww3_strt.F90:
```diff
@@ -816,10 +816,19 @@ PROGRAM W3STRT
+    INXOUT = 'FTCH'
+    
+    CALL NEXTLN ( COMSTR , NDSI , NDSEN )
+    READ (NDSI,*,END=801,ERR=802) FETCH
```
After the recent PR we need to update this to be: 

READ (NDSI,*,IOSTAT=IERR) FETCH 
IF (IERR.NE.0) CALL EXTIOF(NDSE,IERR,'W3STRT','INPUT',11)

JessicaMeixner-NOAA reviewed (commented):
@MHBalsmeier  Apologies again for losing track of this on my end.  We've had some recent PRs that have gone in that are going to require some minor updates.  I have tried to make comments in each place as to what those updates are needed to be.  As soon as you make those updates and merge in develop we'll do one last round of regression tests on our end and then get this merged. Thanks for your patience and sorry again for the delay.

MHBalsmeier committed:
merged in recent changes in develop

MHBalsmeier committed:
implemented changes requested by Jessica Meixner

MHBalsmeier committed:
modifications for ITYPE = 6

after the latest changes to develop, ITYPE = 6 was not functional anymore, so some small changes were necessary to w3iorsmd and ww3_strt

MHBalsmeier committed:
Merge branch 'jonswap_init_fetch_override' of https://github.com/MHBalsmeier/WW3 into jonswap_init_fetch_override

MHBalsmeier commented:
@JessicaMeixner-NOAA No worries, I merged develop back into my branch and implemented the changes you requested. Also I had to do some modifications to keep the new ITYPE = 6 running with the recent changes that were made to develop.
Appended you find the regtest results for the limited subset that I am able to run.

Interestingly, I now keep getting tons of lines

Note: The following floating-point exceptions are signalling: IEEE_UNDERFLOW_FLAG IEEE_DENORMAL

which was not the case before. Maybe this is due to the recent changes to develop.

[matrixCompSummary.txt](https://github.com/user-attachments/files/21437742/matrixCompSummary.txt)
[matrixCompFull.txt](https://github.com/user-attachments/files/21437743/matrixCompFull.txt)

JessicaMeixner-NOAA commented:
@MHBalsmeier - I haven't seen those lines in recent testing.  Is this in compilation? When running certain tests?  I'll keep an eye out for it. 

We did update the debug flags for building?  That could be related. 

I will start running tests now, and should be able to do a final test/review on Monday.  Thanks for the updates!

MHBalsmeier commented:
@JessicaMeixner-NOAA These warnings occur during the execution (of ww3_shel I believe), for example, running

./bin/run_cmake_test ../model/ ww3_tp1.1

leads to
<img width="977" height="338" alt="image" src="https://github.com/user-attachments/assets/67a78e14-2ced-4e01-b50b-78713f2fbcf9" />

JessicaMeixner-NOAA reviewed (approved):
```
**********************************************************************
********************* non-identical cases ****************************
**********************************************************************
mww3_test_03/./work_PR3_UQ_MPI_e                     (1 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (15 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (15 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (12 files differ)
mww3_test_03/./work_PR2_UNO_MPI_e                     (1 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (16 files differ)
mww3_test_03/./work_PR3_UQ_MPI_e_c                     (1 files differ)
mww3_test_03/./work_PR1_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UQ_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UQ_MPI_d2                     (15 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (16 files differ)
mww3_test_03/./work_PR3_UNO_MPI_e                     (1 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (17 files differ)
mww3_test_03/./work_PR3_UNO_MPI_e_c                     (1 files differ)
mww3_test_08/./work_lowres_jonswap                     (directory not found)
mww3_test_09/./work_MPI_ASCII                     (0 files differ)
ww3_tp2.10/./work_MPI_OMPH                     (7 files differ)
ww3_tp2.6/./work_ST4_ASCII                     (0 files differ)
ww3_ufs1.3/./work_a                     (3 files differ)
```

Only new tests which are not in develop are "diff" beyond expected diffs. 
[matrixCompFull.txt](https://github.com/user-attachments/files/21473752/matrixCompFull.txt)
[matrixCompSummary.txt](https://github.com/user-attachments/files/21473753/matrixCompSummary.txt)
[matrixDiff.txt](https://github.com/user-attachments/files/21473754/matrixDiff.txt)
